<a href="https://colab.research.google.com/github/salahagnir/CA2/blob/main/Programming.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
pip install dash

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.2/10.2 MB 22.1 MB/s eta 0:00:00


In [2]:
import requests  # Importing the requests library

baseurl = 'https://rickandmortyapi.com/api/'  # Defining the base URL for the API
endpoint = 'character'  # Defining the specific endpoint for character data

# Function to make an HTTP GET request to the specified endpoint and return the JSON response
def main_requests(baseurl, endpoint):
    r = requests.get(baseurl + endpoint)
    return r.json()

# Making the main request to get data for the 'character' endpoint
data = main_requests(baseurl, endpoint)

# Printing the retrieved data from the API in JSON format
print(data)

{'info': {'count': 826, 'pages': 42, 'next': 'https://rickandmortyapi.com/api/character?page=2', 'prev': None}, 'results': [{'id': 1, 'name': 'Rick Sanchez', 'status': 'Alive', 'species': 'Human', 'type': '', 'gender': 'Male', 'origin': {'name': 'Earth (C-137)', 'url': 'https://rickandmortyapi.com/api/location/1'}, 'location': {'name': 'Citadel of Ricks', 'url': 'https://rickandmortyapi.com/api/location/3'}, 'image': 'https://rickandmortyapi.com/api/character/avatar/1.jpeg', 'episode': ['https://rickandmortyapi.com/api/episode/1', 'https://rickandmortyapi.com/api/episode/2', 'https://rickandmortyapi.com/api/episode/3', 'https://rickandmortyapi.com/api/episode/4', 'https://rickandmortyapi.com/api/episode/5', 'https://rickandmortyapi.com/api/episode/6', 'https://rickandmortyapi.com/api/episode/7', 'https://rickandmortyapi.com/api/episode/8', 'https://rickandmortyapi.com/api/episode/9', 'https://rickandmortyapi.com/api/episode/10', 'https://rickandmortyapi.com/api/episode/11', 'https://ri

In [3]:
import pandas as pd  # Importing pandas for data manipulation

# Function to make an HTTP GET request to the specified endpoint and return the JSON response
def main_requests(baseurl, endpoint):
    r = requests.get(baseurl + endpoint)
    return r.json()

# Function to extract the total number of pages from the API response
def get_pages(response):
    return response['info']['pages']

# Function to parse the JSON response and create a list of dictionaries containing character data (ID, name, and number of episodes)
def parse_json(response):
    charlist = []  # Initializing an empty list to store character data
    for item in response['results']:
        char = {
            'id': item['id'],  # Extracting character ID from the response
            'name': item['name'],  # Extracting character name from the response
            'no_ep': len(item['episode'])  # Determining the number of episodes the character appears in
        }
        charlist.append(char)  # Adding the character's data to the list
    return charlist  # Returning the list of character data

# Replace 'baseurl' and 'endpoint' with your API base URL and endpoint
baseurl = 'https://rickandmortyapi.com/api/'
endpoint = 'character'

data = main_requests(baseurl, endpoint)  # Making the main request to get data for the 'character' endpoint

mainlist = []  # Initializing an empty list to store all character data

# Parsing and adding character data from subsequent pages (if any) to the mainlist
mainlist.extend(parse_json(data))  # Adding character data from the first page to the mainlist

pages = get_pages(data)  # Getting the total number of pages
for x in range(2, pages + 1):  # Looping through subsequent pages starting from page 2
    next_data = main_requests(baseurl, endpoint + f'?page={x}')  # Requesting data for the next page
    mainlist.extend(parse_json(next_data))  # Adding character data from the next page to the mainlist

# Creating a DataFrame from the collected character data and saving it to a CSV file named 'charlist.csv'
df = pd.DataFrame(mainlist)
df.to_csv('charlist.csv', index=False)

This Python code fetches character data from the Rick and Morty API, organizes it into a Pandas DataFrame, and saves it as a CSV file. It's useful for automating data retrieval from APIs, converting JSON responses into structured data, and storing it for analysis or further use in a CSV format.

In [4]:
df

id          name  no_ep
0      1  Rick Sanchez     51
1      2   Morty Smith     51
2      3  Summer Smith     42
3      4    Beth Smith     42
4      5   Jerry Smith     39
..   ...           ...    ...
821  822   Young Jerry      1
822  823    Young Beth      1
823  824    Young Beth      1
824  825   Young Jerry      1
825  826  Butter Robot      1

[826 rows x 3 columns]


The output presents a structured list of characters from the Rick and Morty series with their respective IDs, names, and the number of episodes each character has appeared in. It's useful for data analysis, reference, database input, or documentation purposes due to its organized tabular format.

In [5]:
# Summarizes DataFrame structure.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 826 entries, 0 to 825
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   id      826 non-null    int64 
 1   name    826 non-null    object
 2   no_ep   826 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 19.5+ KB


The output represents a Pandas DataFrame with 826 rows and 3 columns. It shows the column names, data types, and the number of non-null entries in each column. This summary is useful for quickly understanding the DataFrame's structure and memory usage.

In [6]:
# Provides summary statistics for numerical columns in the DataFrame.
df.describe()

id       no_ep
count  826.000000  826.000000
mean   413.500000    1.533898
std    238.589955    3.557587
min      1.000000    1.000000
25%    207.250000    1.000000
50%    413.500000    1.000000
75%    619.750000    1.000000
max    826.000000   51.000000

The statistical summary reveals that character identity numbers ('id') span from 1 to 826, with an average value of approximately 413.5 and a standard deviation of around 238.59. Characters' appearances in episodes ('no_ep') range from 1 to 51, with an average involvement of about 1.53 episodes and a standard deviation of roughly 3.56. The quartiles suggest a diverse distribution of both character IDs and episode appearances among the dataset

In [7]:
# Calculate the percentage of missing values in each column
missing_percentage = (df.isnull().sum() / len(df)) * 100

print("Percentage of missing values in each column:")
print(missing_percentage)

Percentage of missing values in each column:
id       0.0
name     0.0
no_ep    0.0
dtype: float64


There are no missing values (0.0%) in any of the columns ('id', 'name', 'no_ep') of the DataFrame.

In [8]:
import dash
import dash_core_components as dcc
import dash_html_components as html
import plotly.express as px


<ipython-input-8-35744baab968>:2: UserWarning: 
The dash_core_components package is deprecated. Please replace
`import dash_core_components as dcc` with `from dash import dcc`
  import dash_core_components as dcc
<ipython-input-8-35744baab968>:3: UserWarning: 
The dash_html_components package is deprecated. Please replace
`import dash_html_components as html` with `from dash import html`
  import dash_html_components as html


In [9]:
# Importing Plotly Express library as px for interactive visualization capabilities.
import plotly.express as px

In [10]:
# Creating a treemap to visualize the proportion of episodes for each character
fig = px.treemap(df, path=['name'], values='no_ep',
                 title='Proportion of Episodes for Each Character')
fig.show()


The treemap generated using the code displays a visual representation of the proportion of episodes for each character based on the 'name' column and the number of episodes ('no_ep') in the DataFrame.

Morty Smith and Rick Sanchez are depicted with larger sections in the treemap compared to Sck Zack, it indicates that Morty Smith and Rick Sanchez have a higher count of episodes they've appeared in compared to Sck Zack. The larger rectangles representing Morty Smith and Rick Sanchez signify a larger share of episodes attributed to these characters, while the smaller rectangle for Sck Zack indicates a relatively smaller number of episodes involving this character.

This treemap visualization effectively showcases the varying proportions of episode appearances among different characters in the dataset.

In [11]:
# Initialize the Dash app
app = dash.Dash(__name__)

In [12]:
# Define the layout of the app
app.layout = html.Div([
    html.H1("Number of Episodes per Character"),
    dcc.Graph(
        id='episode-bar-chart',
        figure=px.bar(df, x='name', y='no_ep', color='name',
                      labels={'no_ep': 'Number of Episodes', 'name': 'Character'})
    )
])

# Run the app
if __name__ == '__main__':
    app.run_server(debug=True)

<IPython.core.display.Javascript object>

Characters like Rick Sanchez, Amish Cyborg, Baby Wizard, etc., are shown with the highest number of episodes, as they have bars extending farther to the right on the x-axis. Meanwhile, characters like Long Hair Rick and others have smaller bars, indicating fewer episodes associated with them.